### DISCUSSION TITLE
Is it possible to execute multiple commands with diesel

### DISCUSSION THREAD

**Discussion by jiangxiaoqiang** [diesel-rs/diesel#3096] [Q&A]
I couldn't find any information on possibility to execute multiple commands in one statement. is it possible to support this or already support this? I did not found any example about this, for example, execute sql like this:

```
begin;

declare article_cursor cursor for
select
	*,
	count_estimate('select * from article')
from
	(
	select
		"article"."id",
		"article"."user_id",
		"article"."title",
		"article"."author",
		"article"."guid",
		"article"."created_time",
		"article"."updated_time",
		"article"."link",
		"article"."pub_time",
		"article"."sub_source_id",
		"article"."cover_image",
		"article"."channel_reputation",
		"article"."editor_pick"
	from
		"article"
	where
		"article"."id" > $1
	order by
		"article"."created_time" desc
	limit $2 offset $3 ) t;

fetch 10
from
article_cursor;
end;

```

and this is my rust code to generate this sql statement but execute failed in PostgreSQL 13:

```
pub fn handle_big_table_query<T: QueryFragment<Pg>>(this: &PgBigTablePaginated<T>, mut out: AstPass<Pg>)-> QueryResult<()>{
    // https://www.sqlstyle.guide/
    let select_str = format!("BEGIN; DECLARE article_cursor CURSOR FOR SELECT *, count_estimate('select * from {}') FROM  ", this.table_name);
    out.push_sql(select_str.as_str());
    if this.is_sub_query {
        out.push_sql("(");
    }
    this.query.walk_ast(out.reborrow())?;
    if this.is_sub_query {
        out.push_sql(" LIMIT ");
        out.push_bind_param::<BigInt, _>(&this.per_page)?;
        out.push_sql(" OFFSET ");
        let offset = (this.page - 1) * this.per_page;
        out.push_bind_param::<BigInt, _>(&offset)?;
        let fetch_str = format!(") t; FETCH {} FROM article_cursor; END;", this.per_page);
        out.push_sql(fetch_str.as_str());
    }
    Ok(())
}

```

and this is my DDL of article table:

```
CREATE TABLE public.article (
	id int8 NOT NULL GENERATED ALWAYS AS IDENTITY,
	user_id int8 NOT NULL,
	title varchar(512) NOT NULL,
	author varchar(256) NOT NULL,
	guid varchar(512) NOT NULL,
	created_time int8 NOT NULL,
	updated_time int8 NOT NULL,
	link varchar(512) NULL,
	pub_time timestamptz NULL,
	sub_source_id int8 NOT NULL,
	cover_image varchar(1024) NULL,
	channel_reputation int4 NOT NULL DEFAULT 0,
	editor_pick int4 NULL DEFAULT 0,
	CONSTRAINT article_id_seq_pk PRIMARY KEY (id),
	CONSTRAINT article_title_sub_source_id_key UNIQUE (title, sub_source_id)
);


```

and this is the count_estimate function:

```
CREATE OR REPLACE FUNCTION public.count_estimate(query text)
 RETURNS integer
 LANGUAGE plpgsql
AS $function$
DECLARE
    rec   record;
    ROWS  INTEGER;
BEGIN
    FOR rec IN EXECUTE 'EXPLAIN ' || query LOOP
        ROWS := SUBSTRING(rec."QUERY PLAN" FROM ' rows=([[:digit:]]+)');
        EXIT WHEN ROWS IS NOT NULL;
    END LOOP;
 
    RETURN ROWS;
END
$function$
;
;

```

this SQL works fine in the DBeaver, but failed when execute with diesel. The output error message look like this:

```
thread 'rocket-worker-thread' panicked at 'called `Result::unwrap()` on an `Err` value: DatabaseError(__Unknown, "cannot insert multiple commands into a prepared statement")', /usr/local/cargo/git/checkouts/rust_wheel-8476ff1b418e67f8/5036ed5/src/common/util/model_convert.rs:13:32
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

I already read that PostgreSQL support multi commands from here: https://www.postgresql.org/docs/current/libpq-exec.html, but did not know how to enable it with diesel.

**Comment by weiznich** [comment]
Discussion threads opened in the idea category are expected to outline details about a new API. I expect from such a request to present at least the following points:
* A use case with complete details
* How this would fit into diesels existing API
* How this interacts with existing API's
* Which API's/features are potentially broken by the proposed feature
* What are alternatives to the proposed features

You can think of this category as rust RFC process like section. 
It's specifically not a place to post questions for help.  As this is already your second post here without substantial details this is my last warning. I will move this one to the question section, any further idea post that does not provide the details mentioned above will be reported as spam.

To answer your question: No, multi statement queries are not possible using diesel. Diesel uses prepared statements internally and prepared statements just expect one statement per query not more than one. It's not up to discussion to change this.

As additional side note: The pagination example from our guide is specifically **not** an example how to implement a optimized paginate query. It's an example how to extend diesels query builder, that just happens to use an fitting query.

**Reply by jiangxiaoqiang** [reply]
could you tell me why only support prepared statements? you said change this, means it only possible to choose one? it is not possible to support both `PQexec` and `PQexecParams`? when we decided to use PQexecParams, using PQexec will be impossible? sorrry this question maybe stupid,  I am still learning with the rust.

**Reply by weiznich** [reply]
Again: Not using prepared statements is not up to discussion for diesel. We've explicitly chosen to only use prepared statements as everything else will lead to security issues due to sql injections. That's nothing that is related to rust, but that's just that what is considered good practice for interacting with databases.

If you want to issue two statements in diesel you can already do that by just running two separate queries.

**Reply by jiangxiaoqiang** [reply]
ok, I see, is consider about the secure issue and keep best practice, thank you for your patient. I will try other way to do this.
